File: Python/42. Trapping Rain Water.py
class Solution:
    def trap(self, height: list[int]) -> int:
        left = []
        result = 0

        for i in range(len(height)):
            if len(left) > 0 and height[i] >= left[-1][1]:
                # right = [i, height[i]]
                temp = left.pop()[1]
                while len(left) > 0 and left[-1][1] <= height[i]:
                    l = left.pop()
                    result += ((l[1]-temp) * (i-l[0]-1))
                    temp = l[1]
                if len(left) > 0 and left[-1][1] >= height[i]:
                    result += ((height[i] - temp) * (i-left[-1][0]-1))
            left.append([i, height[i]])
            
        return result

File: Javascript/42. Trapping Rain Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let left = Array()
    
    let n = height.length
    let result = 0

    for (let i = 0; i < n; i++) {
        if (left.length > 0 && height[i] >= left[left.length-1][1]) {
            let right = [i, height[i]]
            let temp = left.pop()[1]
            while (left.length && left[left.length-1][1] <= right[1]) {
                let l = left.pop()
                result += ( (l[1]-temp) * (i-l[0]-1) )
                temp = l[1]
            }
            if (left.length > 0 && left[left.length-1][1] >= right[1]) {
                result += ( (right[1]-temp) * (right[0]-left[left.length-1][0]-1) )
            }
        }
        left.push([i, height[i]])
    }
    return result
};